kcmdline.c

```diff
@@ -0,0 +1,89 @@
+#include "kcmdline.h"
+
+static const CHAR8 allowed[6][12] = {
+	"^console=",
+	"^root=oci:",
+	"root=atomix",
+	"ro",
+	"quiet",
+	"verbose"
+};
+
+#define MAX_TOKENS 128
+
+BOOLEAN is_allowed(const CHAR8 *input) {
+	int len = 0;
+	const CHAR8 * token;
+    int num_allowed = sizeof(allowed) / sizeof(allowed[0]);
+	for (int i=0; i<num_allowed; i++) {
+		token = allowed[i];
+		len = strlena(token);
+		if (token[0] == '^') {
+			len = len - 1;
+			token = &token[1];
+		}
+		if (strncmpa(token, input, len) == 0) {
+			return TRUE;
+		}
+	}
+	return FALSE;
+}
+
+// check cmdline to make sure it contains only allowed words.
+// return EFI_SUCCESS on safe, EFI_SECURITY_VIOLATION on unsafe;
+EFI_STATUS check_cmdline(CONST CHAR8 *cmdline, UINTN cmdline_len) {
+	CHAR8 c = '\0';
+	CHAR8 *buf = NULL;
+	CHAR8 *tokens[MAX_TOKENS];
+	int i;
+	int start = -1;
+	int num_toks = 0;
+	buf = AllocatePool(cmdline_len);
```

Comment:
Is it an EFI thing that we don't worry about freeing this memory?  Will it get freed upon stubby executing the kernel anyway?

Reply:
that was my thought.

Reply:
Lacking a more definitive reference...  https://edk2-docs.gitbook.io/edk-ii-uefi-driver-writer-s-guide/5_uefi_services/51_services_that_uefi_drivers_commonly_use/511_memory_allocation_services doesn't say.  Freeing seems best.

Reply:
I can call FreePool, because yuo're right that it "sounds best", but am I missing something?

This is a program "kernel.efi".  It allocates some memory, and then either:
 a.) executes linux : linux doesn't care or know about whatever memory was allocated in EFI, does it?
 b.) exits failure: i assumed that uefi runtime will take back any memory by the process.

I'll add FreePool call, but I'm just asking if I understand things right.

Reply:
updated. used freepool